# Name:- Parshwa Shah
# Experiment No.:- 7
# Roll No.- 34
# UID:- 2019230071
# Batch:- B

<h2>Aim:- To perform text analysis and email spam classification </h2>

<h3>Importing the libraries </h3>

In [1]:
import pandas as pd

<h3>Downloading the dataset </h3>

In [2]:
! gdown --id 1Bgkj-BUL9Hjv6zzQAhAiD5CIrG5yqKwo

Downloading...
From: https://drive.google.com/uc?id=1Bgkj-BUL9Hjv6zzQAhAiD5CIrG5yqKwo
To: /content/emails.csv
100% 8.95M/8.95M [00:00<00:00, 199MB/s]


<h3>Importing the dataset </h3>

In [3]:
dataset = pd.read_csv('./emails.csv')

In [4]:
dataset

text  spam
0     Subject: naturally irresistible your corporate...     1
1     Subject: the stock trading gunslinger  fanny i...     1
2     Subject: unbelievable new homes made easy  im ...     1
3     Subject: 4 color printing special  request add...     1
4     Subject: do not have money , get software cds ...     1
...                                                 ...   ...
5723  Subject: re : research and development charges...     0
5724  Subject: re : receipts from visit  jim ,  than...     0
5725  Subject: re : enron case study update  wow ! a...     0
5726  Subject: re : interest  david ,  please , call...     0
5727  Subject: news : aurora 5 . 2 update  aurora ve...     0

[5728 rows x 2 columns]

How many emails are in the dataset?

5728

In [5]:
dataset['spam'].value_counts()

0    4360
1    1368
Name: spam, dtype: int64

How many of the emails are spam?

1368 are spam

<h3>Check for null values in dataset </h3>

In [6]:
dataset.isna().sum()

text    0
spam    0
dtype: int64

<h3>Check the most common beginning word in emails </h3>

In [7]:
dataset['startword'] = dataset.apply(lambda x: x['text'].split()[0],axis=1)

In [8]:
dataset

text  spam startword
0     Subject: naturally irresistible your corporate...     1  Subject:
1     Subject: the stock trading gunslinger  fanny i...     1  Subject:
2     Subject: unbelievable new homes made easy  im ...     1  Subject:
3     Subject: 4 color printing special  request add...     1  Subject:
4     Subject: do not have money , get software cds ...     1  Subject:
...                                                 ...   ...       ...
5723  Subject: re : research and development charges...     0  Subject:
5724  Subject: re : receipts from visit  jim ,  than...     0  Subject:
5725  Subject: re : enron case study update  wow ! a...     0  Subject:
5726  Subject: re : interest  david ,  please , call...     0  Subject:
5727  Subject: news : aurora 5 . 2 update  aurora ve...     0  Subject:

[5728 rows x 3 columns]

In [9]:
dataset['startword'].value_counts()

Subject:    5728
Name: startword, dtype: int64

Could a spam classifier potentially benefit from including the frequency of the word that appears in every email?

Yes -- the number of times the word appears might help us differentiate spam from ham.

<h3>Check the longest email </h3>

In [10]:
import numpy as np
lengths = dataset["text"].str.len()
argmax = np.where(lengths == lengths.max())[0]
dataset.iloc[argmax]

text  spam startword
2650  Subject: from the enron india newsdesk - april...     0  Subject:

In [11]:
longesttext = dataset.iloc[argmax].to_numpy().ravel().tolist()

In [12]:
len(longesttext[0])

43952

<h3>Delete the startword column </h3>

In [13]:
del dataset['startword']

In [14]:
dataset

text  spam
0     Subject: naturally irresistible your corporate...     1
1     Subject: the stock trading gunslinger  fanny i...     1
2     Subject: unbelievable new homes made easy  im ...     1
3     Subject: 4 color printing special  request add...     1
4     Subject: do not have money , get software cds ...     1
...                                                 ...   ...
5723  Subject: re : research and development charges...     0
5724  Subject: re : receipts from visit  jim ,  than...     0
5725  Subject: re : enron case study update  wow ! a...     0
5726  Subject: re : interest  david ,  please , call...     0
5727  Subject: news : aurora 5 . 2 update  aurora ve...     0

[5728 rows x 2 columns]

<h3>Install nltk library </h3>

In [15]:
! pip install nltk

<h3>Make the email message lowercase </h3>

In [16]:
import re
def lowercase(textvalue):
  textvalue= re.sub(r"[^a-zA-Z0-9]", " ", textvalue.lower())
  return textvalue

In [17]:
dataset['text'] = dataset['text'].apply(lambda x: lowercase(x))

In [18]:
dataset

text  spam
0     subject  naturally irresistible your corporate...     1
1     subject  the stock trading gunslinger  fanny i...     1
2     subject  unbelievable new homes made easy  im ...     1
3     subject  4 color printing special  request add...     1
4     subject  do not have money   get software cds ...     1
...                                                 ...   ...
5723  subject  re   research and development charges...     0
5724  subject  re   receipts from visit  jim    than...     0
5725  subject  re   enron case study update  wow   a...     0
5726  subject  re   interest  david    please   call...     0
5727  subject  news   aurora 5   2 update  aurora ve...     0

[5728 rows x 2 columns]

<h3>Remove punctuation from dataset </h3>

In [19]:
def removePunc(text):
  text = re.sub(r'[^\w\s]','',text)
  return text

In [20]:
dataset['text'] = dataset['text'].apply(lambda x: removePunc(x))

In [21]:
dataset

text  spam
0     subject  naturally irresistible your corporate...     1
1     subject  the stock trading gunslinger  fanny i...     1
2     subject  unbelievable new homes made easy  im ...     1
3     subject  4 color printing special  request add...     1
4     subject  do not have money   get software cds ...     1
...                                                 ...   ...
5723  subject  re   research and development charges...     0
5724  subject  re   receipts from visit  jim    than...     0
5725  subject  re   enron case study update  wow   a...     0
5726  subject  re   interest  david    please   call...     0
5727  subject  news   aurora 5   2 update  aurora ve...     0

[5728 rows x 2 columns]

<h3>Remove stopwords from emails </h3>

In [22]:
from nltk.corpus import stopwords

In [23]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [24]:
stopwords = stopwords.words("english")
len(stopwords)

179

In [25]:
def remove_stopwords(text,stopwords):
  words=text.split()
  words = [w for w in words if w not in stopwords]
  return words

In [26]:
dataset['text'] = dataset['text'].apply(lambda x: remove_stopwords(x,stopwords))

In [27]:
dataset

text  spam
0     [subject, naturally, irresistible, corporate, ...     1
1     [subject, stock, trading, gunslinger, fanny, m...     1
2     [subject, unbelievable, new, homes, made, easy...     1
3     [subject, 4, color, printing, special, request...     1
4     [subject, money, get, software, cds, software,...     1
...                                                 ...   ...
5723  [subject, research, development, charges, gpg,...     0
5724  [subject, receipts, visit, jim, thanks, invita...     0
5725  [subject, enron, case, study, update, wow, day...     0
5726  [subject, interest, david, please, call, shirl...     0
5727  [subject, news, aurora, 5, 2, update, aurora, ...     0

[5728 rows x 2 columns]

<h3>Perform stemming on emails </h3>

In [28]:
from nltk.stem.porter import PorterStemmer
def stemming(words):
  stemmed = [PorterStemmer().stem(w) for w in words]
  return stemmed

In [29]:
dataset['text'] = dataset['text'].apply(lambda x: stemming(x))

In [30]:
dataset

text  spam
0     [subject, natur, irresist, corpor, ident, lt, ...     1
1     [subject, stock, trade, gunsling, fanni, merri...     1
2     [subject, unbeliev, new, home, made, easi, im,...     1
3     [subject, 4, color, print, special, request, a...     1
4     [subject, money, get, softwar, cd, softwar, co...     1
...                                                 ...   ...
5723  [subject, research, develop, charg, gpg, forwa...     0
5724  [subject, receipt, visit, jim, thank, invit, v...     0
5725  [subject, enron, case, studi, updat, wow, day,...     0
5726  [subject, interest, david, pleas, call, shirle...     0
5727  [subject, news, aurora, 5, 2, updat, aurora, v...     0

[5728 rows x 2 columns]

<h3>Convert email list to string </h3>

In [31]:
def join_string(x):
  return ' '.join(x)

In [32]:
dataset['text'] = dataset['text'].apply(lambda x: join_string(x))

In [33]:
dataset

text  spam
0     subject natur irresist corpor ident lt realli ...     1
1     subject stock trade gunsling fanni merril muzo...     1
2     subject unbeliev new home made easi im want sh...     1
3     subject 4 color print special request addit in...     1
4     subject money get softwar cd softwar compat gr...     1
...                                                 ...   ...
5723  subject research develop charg gpg forward shi...     0
5724  subject receipt visit jim thank invit visit ls...     0
5725  subject enron case studi updat wow day super t...     0
5726  subject interest david pleas call shirley cren...     0
5727  subject news aurora 5 2 updat aurora version 5...     0

[5728 rows x 2 columns]

<h3>Apply CountVectorizer to vectorize emails </h3>

In [34]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

In [35]:
vec = CountVectorizer()
X = vec.fit_transform(dataset["text"])

In [36]:
corpus = pd.DataFrame(X.toarray(), columns=vec.get_feature_names())
corpus.head()

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  0000  000000  00000000  0000000000  000000000003619  \
0   0    0     0       0         0           0                0   
1   0    0     0       0         0           0                0   
2   0    0     0       0         0           0                0   
3   0    0     0       0         0           0                0   
4   0    0     0       0         0           0                0   

   000000000003991  000000000003997  000000000005168  ...  zwzm  zxghlajf  \
0                0                0                0  ...     0         0   
1                0                0                0  ...     0         0   
2                0                0                0  ...     0         0   
3                0                0                0  ...     0         0   
4                0                0                0  ...     0         0   

   zyban  zyc  zygoma  zymg  zzmacmac  zzn  zzncacst  zzzz  
0      0    0       0     0         0    0         0     0  
1      0    0       0     0         0    0         0     0  
2      0    0       0     0         0    0         0     0  
3      0    0       0     0         0    0         0     0  
4      0    0       0     0         0    0         0     0  

[5 rows x 29222 columns]

<h3>Create Document Term Matrix of emails </h3>

In [37]:
corpus_dtm = corpus.copy(deep=True)

In [38]:
corpus_dtm

00  000  0000  000000  00000000  0000000000  000000000003619  \
0      0    0     0       0         0           0                0   
1      0    0     0       0         0           0                0   
2      0    0     0       0         0           0                0   
3      0    0     0       0         0           0                0   
4      0    0     0       0         0           0                0   
...   ..  ...   ...     ...       ...         ...              ...   
5723   0    0     0       0         0           0                0   
5724   1    0     0       0         0           0                0   
5725   4    0     0       0         0           0                0   
5726   0    0     0       0         0           0                0   
5727   0    0     0       0         0           0                0   

      000000000003991  000000000003997  000000000005168  ...  zwzm  zxghlajf  \
0                   0                0                0  ...     0         0   
1                   0                0                0  ...     0         0   
2                   0                0                0  ...     0         0   
3                   0                0                0  ...     0         0   
4                   0                0                0  ...     0         0   
...               ...              ...              ...  ...   ...       ...   
5723                0                0                0  ...     0         0   
5724                0                0                0  ...     0         0   
5725                0                0                0  ...     0         0   
5726                0                0                0  ...     0         0   
5727                0                0                0  ...     0         0   

      zyban  zyc  zygoma  zymg  zzmacmac  zzn  zzncacst  zzzz  
0         0    0       0     0         0    0         0     0  
1         0    0       0     0         0    0         0     0  
2         0    0       0     0         0    0         0     0  
3         0    0       0     0         0    0         0     0  
4         0    0       0     0         0    0         0     0  
...     ...  ...     ...   ...       ...  ...       ...   ...  
5723      0    0       0     0         0    0         0     0  
5724      0    0       0     0         0    0         0     0  
5725      0    0       0     0         0    0         0     0  
5726      0    0       0     0         0    0         0     0  
5727      0    0       0     0         0    0         0     0  

[5728 rows x 29222 columns]

How many terms are in dtm?

29222

<h3> Make spdtm by taking words which are present in atleast 5%of emails </h3>

In [39]:
vec2 = CountVectorizer(min_df=0.05)
X2 = vec2.fit_transform(dataset["text"])

In [40]:
corpus2 = pd.DataFrame(X2.toarray(), columns=vec2.get_feature_names())
corpus2.head()

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  01  02  03  04  05  06  07  08  ...  week  well  wish  within  \
0   0    0   0   0   0   0   0   0   0   0  ...     0     0     0       1   
1   0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   
2   0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   
3   0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   
4   0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   

   without  work  would  write  www  year  
0        1     0      0      0    0     0  
1        0     0      0      0    0     0  
2        0     0      0      0    0     0  
3        0     0      0      0    0     0  
4        0     0      0      0    0     0  

[5 rows x 366 columns]

In [41]:
corpus_spdtm = corpus2.copy(deep=True)
corpus_spdtm

00  000  01  02  03  04  05  06  07  08  ...  week  well  wish  within  \
0      0    0   0   0   0   0   0   0   0   0  ...     0     0     0       1   
1      0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   
2      0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   
3      0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   
4      0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   
...   ..  ...  ..  ..  ..  ..  ..  ..  ..  ..  ...   ...   ...   ...     ...   
5723   0    0   0   1   3   0   0   0   1   5  ...     0     0     0       0   
5724   1    0   0   1   0   0   0   0   1   1  ...     0     0     0       0   
5725   4    0   1   0   0   2   0   1   0   0  ...     1     1     0       0   
5726   0    0   0   0   0   0   2   4   0   0  ...     0     0     0       0   
5727   0    0   0   0   0   0   0   0   0   0  ...     0     0     0       0   

      without  work  would  write  www  year  
0           1     0      0      0    0     0  
1           0     0      0      0    0     0  
2           0     0      0      0    0     0  
3           0     0      0      0    0     0  
4           0     0      0      0    0     0  
...       ...   ...    ...    ...  ...   ...  
5723        0     0      1      0    0     0  
5724        0     0      0      0    0     0  
5725        0     1      0      1    0     0  
5726        0     0      0      0    0     0  
5727        0     0      0      0    2     1  

[5728 rows x 366 columns]

How many terms are in spdtm?

366

In [42]:
corpus_dtm

00  000  0000  000000  00000000  0000000000  000000000003619  \
0      0    0     0       0         0           0                0   
1      0    0     0       0         0           0                0   
2      0    0     0       0         0           0                0   
3      0    0     0       0         0           0                0   
4      0    0     0       0         0           0                0   
...   ..  ...   ...     ...       ...         ...              ...   
5723   0    0     0       0         0           0                0   
5724   1    0     0       0         0           0                0   
5725   4    0     0       0         0           0                0   
5726   0    0     0       0         0           0                0   
5727   0    0     0       0         0           0                0   

      000000000003991  000000000003997  000000000005168  ...  zwzm  zxghlajf  \
0                   0                0                0  ...     0         0   
1                   0                0                0  ...     0         0   
2                   0                0                0  ...     0         0   
3                   0                0                0  ...     0         0   
4                   0                0                0  ...     0         0   
...               ...              ...              ...  ...   ...       ...   
5723                0                0                0  ...     0         0   
5724                0                0                0  ...     0         0   
5725                0                0                0  ...     0         0   
5726                0                0                0  ...     0         0   
5727                0                0                0  ...     0         0   

      zyban  zyc  zygoma  zymg  zzmacmac  zzn  zzncacst  zzzz  
0         0    0       0     0         0    0         0     0  
1         0    0       0     0         0    0         0     0  
2         0    0       0     0         0    0         0     0  
3         0    0       0     0         0    0         0     0  
4         0    0       0     0         0    0         0     0  
...     ...  ...     ...   ...       ...  ...       ...   ...  
5723      0    0       0     0         0    0         0     0  
5724      0    0       0     0         0    0         0     0  
5725      0    0       0     0         0    0         0     0  
5726      0    0       0     0         0    0         0     0  
5727      0    0       0     0         0    0         0     0  

[5728 rows x 29222 columns]

<h3>Apply TF-IDF transformer on DTM </h3>

In [43]:
from sklearn.feature_extraction.text import TfidfTransformer
tfidf_transformer = TfidfTransformer().fit(corpus_dtm)
X_tfidf = tfidf_transformer.transform(corpus_dtm)
print(X_tfidf.shape)

(5728, 29222)


<h3>Get X and y for training model </h3>

In [44]:
X_tfidf

<5728x29222 sparse matrix of type '<class 'numpy.float64'>'
	with 519560 stored elements in Compressed Sparse Row format>

In [45]:
y = dataset["spam"]

In [46]:
y

0       1
1       1
2       1
3       1
4       1
       ..
5723    0
5724    0
5725    0
5726    0
5727    0
Name: spam, Length: 5728, dtype: int64

<h3>Split the dataset into 70% training set and 30% testing set </h3>

In [47]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_tfidf, y, test_size=0.3, random_state=123)

In [48]:
X_train.shape

(4009, 29222)

In [49]:
y_test

3575    0
3964    0
2411    0
1119    1
3278    0
       ..
831     1
65      1
3559    0
280     1
4023    0
Name: spam, Length: 1719, dtype: int64

In [50]:
X_test

<1719x29222 sparse matrix of type '<class 'numpy.float64'>'
	with 153996 stored elements in Compressed Sparse Row format>

<h3>Apply RandomForest Classifier on training set </h3>

In [51]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(random_state=123)
classifier.fit(X_train, y_train)

RandomForestClassifier(random_state=123)

<h3>Getting the results of RF classifier on testing set</h3>

In [52]:
ypred = classifier.predict(X_test)

In [53]:
ypred

array([0, 0, 0, ..., 0, 1, 0])

In [54]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,ypred))

0.9773123909249564


In [55]:
from sklearn.metrics import classification_report
print(classification_report(y_test,ypred))

              precision    recall  f1-score   support

           0       0.97      1.00      0.99      1319
           1       0.99      0.91      0.95       400

    accuracy                           0.98      1719
   macro avg       0.98      0.95      0.97      1719
weighted avg       0.98      0.98      0.98      1719



<h3>Apply Decision Tree Classifier on training set </h3>

In [56]:
from sklearn.tree import DecisionTreeClassifier
spamcart = DecisionTreeClassifier(random_state=123)
spamcart.fit(X_train, y_train)

DecisionTreeClassifier(random_state=123)

<h3>Getting the results of Decision Tree classifier on testing set </h3>

In [57]:
ypred = spamcart.predict(X_test)

In [58]:
ypred

array([0, 0, 0, ..., 0, 1, 0])

In [59]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,ypred))

0.9656777196044212


In [60]:
from sklearn.metrics import classification_report
print(classification_report(y_test,ypred))

              precision    recall  f1-score   support

           0       0.98      0.98      0.98      1319
           1       0.93      0.92      0.93       400

    accuracy                           0.97      1719
   macro avg       0.95      0.95      0.95      1719
weighted avg       0.97      0.97      0.97      1719



<h3>Getting the results of Decision Tree classifier on training set </h3>

In [61]:
trainres = spamcart.predict(X_train)

In [62]:
train_accuracy = classification_report(y_train,trainres)
print(train_accuracy)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      3041
           1       1.00      1.00      1.00       968

    accuracy                           1.00      4009
   macro avg       1.00      1.00      1.00      4009
weighted avg       1.00      1.00      1.00      4009



What is the training set accuracy of spamCART, using a threshold of 0.5 for predictions? 

100%

<h3>Getting the results of Random Forest classifier on training set </h3>

In [63]:
trainres_rf = classifier.predict(X_train)

In [64]:
train_accuracy_rf = classification_report(y_train,trainres_rf)
print(train_accuracy)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      3041
           1       1.00      1.00      1.00       968

    accuracy                           1.00      4009
   macro avg       1.00      1.00      1.00      4009
weighted avg       1.00      1.00      1.00      4009



<h3> Calculate AUC Score on training set on both classifiers </h3>

In [65]:
from sklearn.metrics import roc_auc_score

y_pred_proba = classifier.predict_proba(X_train)[::,1]

auc = roc_auc_score(y_train, y_pred_proba)

print(auc)

1.0


In [66]:
from sklearn.metrics import roc_auc_score
#use model to predict probability that given y value is 1
y_pred_proba = spamcart.predict_proba(X_train)[::,1]

#calculate AUC of model
auc = roc_auc_score(y_train, y_pred_proba)

#print AUC score
print(auc)

1.0


What is the training set AUC of spamCART?

1.0

What is the training set AUC of spamRF?

1.0

<h3> Calculate AUC Score on testing set on both classifiers </h3>

In [67]:
from sklearn.metrics import roc_auc_score
#use model to predict probability that given y value is 1
y_pred_proba = classifier.predict_proba(X_test)[::,1]

#calculate AUC of model
auc = roc_auc_score(y_test, y_pred_proba)

#print AUC score
print(auc)

0.998625852918878


In [68]:
from sklearn.metrics import roc_auc_score
#use model to predict probability that given y value is 1
y_pred_proba = spamcart.predict_proba(X_test)[::,1]

#calculate AUC of model
auc = roc_auc_score(y_test, y_pred_proba)

#print AUC score
print(auc)

0.9506358984078848


What is the testing set accuracy of spamCART, using a threshold of 0.5 for predictions?

97%

What is the testing set AUC of spamCART?

0.9506358984078848

What is the testing set accuracy of spamRF, using a threshold of 0.5 for predictions?

98%

What is the testing set AUC of spamRF?

0.998625852918878

Which model had the best testing set performance, in terms of accuracy and AUC? 

Random Forest performed the best on test set in terms of accuracy and AUC

Comparison between models

Test Set Accuracy: 

1. RandomForest : 98%
2. spamCART : 97%

<h3> Conclusion: Hence, from this experiment, I understood text processing and analysis. Also, I performed email spam classification by text processing, apply vectorizer, training on spamCART and random forest models and got best test accuracy of 98%. This product can be used to classify email spam to save people from fraud. </h3>